### PR TITLE
SEditor PR

### PR DESCRIPTION
**Pull request by hnyu**
This PR moves the SEditor code in https://github.com/hnyu/seditor officially to ALF. The seditor_algorithm.py file has been added with detailed comments and docstrings. 

Verified that seditor_safety_gym_conf.py runs fine. Haven't got time to test if the paper results can be reproduced or not.

### CODE REVIEW

File: alf/algorithms/seditor_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+        "output"  # final output after action editing
+    ],
+    default_value=())
+
+Mode = Enum('AlgorithmMode', ('predict', 'rollout', 'train'))
+
+
+@alf.configurable
+class SEditorAlgorithm(OffPolicyAlgorithm):
+    """Implements the Safe Editor Policy algorithm in the paper:
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: prefix doc-string with r

File: alf/algorithms/seditor_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+
+        self._train_alpha = train_alpha
+        self._soft_clipping = soft_clipping
+
+        self._actor_network = actor_network
+        self._d_actor_network = d_actor_network
+
+        # Here we make a copy of ``actor_network`` and ``d_actor_network`` but
+        # refrain from training them, so that it's easier for us to compute
+        # dqd\hat{a} and dqd\delta a without being interfered by the other network
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: a without => without

File: alf/algorithms/seditor_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+        "da_dist",  # delta distribution
+        "output"  # final output after action editing
+    ],
+    default_value=())
+
+Mode = Enum('AlgorithmMode', ('predict', 'rollout', 'train'))
+
+
+@alf.configurable
+class SEditorAlgorithm(OffPolicyAlgorithm):
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py:86]: you may also update README.md to include this algorithm

File: alf/algorithms/seditor_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+        critic_networks = _make_parallel(critic_network)
+
+        return critic_networks, actor_network, d_actor_network
+
+    def _forward(self,
+                 time_step: TimeStep,
+                 epsilon_greedy: float,
+                 mode: Mode,
+                 opt_ignore: str = 'none') -> ActPredOutput:
+        """
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: r

File: alf/algorithms/seditor_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+            (dqda + action).detach(), action)
+        return loss.sum(list(range(1, loss.ndim)))
+
+    def _actor_train_step(self, inputs, ap_out, a_entropy):
+        """Train UM policy according to Eq 7a, to maximize the utility Q.
+        """
+        alpha = self._log_alpha[0].exp().detach()
+
+        critics = self._compute_critics(self._critic_networks,
+                                        inputs.observation, ap_out.output[0])
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: `ap_out.output[0], ap_out.output[1], ap_out.output[2]` makes the code hard to understand. It's better to use some meaning name other than 0, 1, 2.

File: alf/algorithms/seditor_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+            change_a_loss = a_l2
+        else:
+            with torch.no_grad():
+                a_critics = self._compute_critics(self._critic_networks,
+                                                  inputs.observation,
+                                                  ap_out.output[2])
+            change_a_loss = torch.relu(a_critics[..., 0] - critics[..., 0])
+
+        q2 = torch.stack((-change_a_loss, q), dim=-1)
+        q2 = torch.einsum("bd,d->b", q2, self.reward_weights).sum()
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: Is this (before sum) simply a matrix vector product? `torch.matmul(q2, self.reward_weights)` or simply `q2 @ self.reward_weights` should work

File: alf/algorithms/seditor_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+            output=ap_out.output, state=state, info={
+                'ap_out': ap_out,
+                **imgs
+            })
+
+    def rollout_step(self, inputs: TimeStep, state):
+        ap_out = self._predict_action(inputs, state, mode=Mode.rollout)
+        return AlgStep(output=ap_out.output, state=state, info=ap_out)
+
+    def train_step(self, inputs: TimeStep, state, rollout_info: SEditorInfo):
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: type of rollout_info should be `ActPredOutput`

File: alf/algorithms/seditor_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+        q2 = torch.stack((-change_a_loss, q), dim=-1)
+        q2 = torch.einsum("bd,d->b", q2, self.reward_weights).sum()
+        dqda = nest_utils.grad(ap_out.da, q2)
+
+        actor_loss = self._actor_loss_fn(dqda, ap_out.da)
+        actor_loss -= alpha * da_entropy
+        return actor_loss, change_a_loss, a_l2
+
+    def _critic_train_step(self, inputs: TimeStep, ap_out,
+                           rollout_info: SEditorInfo):
```
Comment (emailweixu) [inline alf/algorithms/seditor_algorithm.py]: the type of rollout_info should be `ActPredOutput`

File: alf/algorithms/seditor_algorithm.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/seditor_algorithm.py]: > a without => without

'a' is together with '\delta'. I've added ":math:" around the math notations to make it less ambiguous.

File: alf/algorithms/seditor_algorithm.py (hunk 5 above)
Comment (hnyu) [reply alf/algorithms/seditor_algorithm.py]: > `ap_out.output[0], ap_out.output[1], ap_out.output[2]` makes the code hard to understand. It's better to use some meaning name other than 0, 1, 2.

updated

File: alf/algorithms/seditor_algorithm.py (hunk 6 above)
Comment (hnyu) [reply alf/algorithms/seditor_algorithm.py]: > matmul

changed

**Comment by hnyu** [comment]
Used hobot's compare_training.py and made sure the change does't change the training.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**